File: src/lib/stores/localSearch.ts
/**
 * Local Search Store - Manages FTS5-powered local message search
 */
import { writable, derived, get } from 'svelte/store';
import { invoke } from '@tauri-apps/api/core';

export interface SearchResult {
    message_id: string;
    channel_id: string;
    server_id: string | null;
    author_id: string;
    author_name: string;
    content: string;
    snippet: string;
    timestamp: number;
    relevance: number;
    has_attachments: boolean;
    has_embeds: boolean;
    is_pinned: boolean;
}

export interface SearchOptions {
    query: string;
    channel_ids?: string[];
    server_ids?: string[];
    author_ids?: string[];
    from_date?: number;
    to_date?: number;
    has_attachments?: boolean;
    has_embeds?: boolean;
    is_pinned?: boolean;
    limit?: number;
    offset?: number;
    sort_by?: 'relevance' | 'date';
    sort_order?: 'asc' | 'desc';
}

export interface IndexableMessage {
    message_id: string;
    channel_id: string;
    server_id: string | null;
    author_id: string;
    author_name: string;
    content: string;
    timestamp: number;
    has_attachments: boolean;
    has_embeds: boolean;
    is_pinned: boolean;
    is_edited: boolean;
}

export interface IndexStats {
    total_messages: number;
    total_channels: number;
    total_servers: number;
    index_size_bytes: number;
    last_indexed_at: number | null;
    oldest_message: number | null;
    newest_message: number | null;
}

// Store state
interface LocalSearchState {
    isOpen: boolean;
    isSearching: boolean;
    isIndexing: boolean;
    query: string;
    results: SearchResult[];
    stats: IndexStats | null;
    error: string | null;
    options: Partial<SearchOptions>;
}

const initialState: LocalSearchState = {
    isOpen: false,
    isSearching: false,
    isIndexing: false,
    query: '',
    results: [],
    stats: null,
    error: null,
    options: {
        sort_by: 'relevance',
        sort_order: 'desc',
        limit: 50,
    },
};

function createLocalSearchStore() {
    const { subscribe, set, update } = writable<LocalSearchState>(initialState);

    return {
        subscribe,
        
        // UI state
        open: () => update(s => ({ ...s, isOpen: true })),
        close: () => update(s => ({ ...s, isOpen: false })),
        toggle: () => update(s => ({ ...s, isOpen: !s.isOpen })),
        
        setQuery: (query: string) => update(s => ({ ...s, query })),
        
        setOptions: (options: Partial<SearchOptions>) => update(s => ({
            ...s,
            options: { ...s.options, ...options },
        })),
        
        clearFilters: () => update(s => ({
            ...s,
            options: {
                sort_by: 'relevance',
                sort_order: 'desc',
                limit: 50,
            },
        })),

        // Search operations
        search: async (query?: string) => {
            update(s => ({
                ...s,
                isSearching: true,
                error: null,
                query: query ?? s.query,
            }));

            try {
                const state = get({ subscribe });
                const searchQuery = query ?? state.query;
                
                if (!searchQuery.trim()) {
                    update(s => ({ ...s, results: [], isSearching: false }));
                    return [];
                }

                const options: SearchOptions = {
                    query: searchQuery,
                    ...state.options,
                };

                const results = await invoke<SearchResult[]>('search_messages', { options });
                update(s => ({ ...s, results, isSearching: false }));
                return results;
            } catch (error) {
                const errorMessage = error instanceof Error ? error.message : String(error);
                update(s => ({ ...s, error: errorMessage, isSearching: false }));
                throw error;
            }
        },

        // Index operations
        indexMessage: async (message: IndexableMessage) => {
            try {
                await invoke('index_message', { message });
            } catch (error) {
                console.error('Failed to index message:', error);
                throw error;
            }
        },

        indexMessages: async (messages: IndexableMessage[]) => {
            update(s => ({ ...s, isIndexing: true }));
            try {
                const count = await invoke<number>('index_messages_batch', { messages });
                update(s => ({ ...s, isIndexing: false }));
                return count;
            } catch (error) {
                update(s => ({ ...s, isIndexing: false }));
                throw error;
            }
        },

        deleteMessage: async (messageId: string) => {
            try {
                await invoke('delete_indexed_message', { messageId });
            } catch (error) {
                console.error('Failed to delete indexed message:', error);
                throw error;
            }
        },

        // Stats and maintenance
        loadStats: async () => {
            try {
                const stats = await invoke<IndexStats>('get_search_stats');
                update(s => ({ ...s, stats }));
                return stats;
            } catch (error) {
                console.error('Failed to load search stats:', error);
                throw error;
            }
        },

        optimize: async () => {
            update(s => ({ ...s, isIndexing: true }));
            try {
                await invoke('optimize_search_index');
                update(s => ({ ...s, isIndexing: false }));
            } catch (error) {
                update(s => ({ ...s, isIndexing: false }));
                throw error;
            }
        },

        clear: async () => {
            update(s => ({ ...s, isIndexing: true }));
            try {
                await invoke('clear_search_index');
                update(s => ({ ...s, results: [], stats: null, isIndexing: false }));
            } catch (error) {
                update(s => ({ ...s, isIndexing: false }));
                throw error;
            }
        },

        reset: () => set(initialState),
    };
}

export const localSearch = createLocalSearchStore();

// Derived stores
export const isSearchOpen = derived(localSearch, $s => $s.isOpen);
export const searchResults = derived(localSearch, $s => $s.results);
export const searchStats = derived(localSearch, $s => $s.stats);
export const isSearchBusy = derived(localSearch, $s => $s.isSearching || $s.isIndexing);

// Helper to auto-index messages from WebSocket/API events
export function createMessageIndexer() {
    const pendingMessages: IndexableMessage[] = [];
    let flushTimeout: ReturnType<typeof setTimeout> | null = null;
    const BATCH_SIZE = 50;
    const FLUSH_DELAY = 2000;

    const flush = async () => {
        if (pendingMessages.length === 0) return;
        
        const batch = pendingMessages.splice(0, BATCH_SIZE);
        try {
            await localSearch.indexMessages(batch);
        } catch (error) {
            console.error('Failed to flush message batch to search index:', error);
            // Re-queue failed messages
            pendingMessages.unshift(...batch);
        }

        // Continue if there are more
        if (pendingMessages.length > 0) {
            flushTimeout = setTimeout(flush, FLUSH_DELAY);
        }
    };

    return {
        queue: (message: IndexableMessage) => {
            pendingMessages.push(message);
            
            // Auto-flush when batch size reached
            if (pendingMessages.length >= BATCH_SIZE) {
                if (flushTimeout) {
                    clearTimeout(flushTimeout);
                }
                flush();
            } else if (!flushTimeout) {
                // Schedule delayed flush for small batches
                flushTimeout = setTimeout(flush, FLUSH_DELAY);
            }
        },
        
        forceFlush: flush,
        
        clear: () => {
            pendingMessages.length = 0;
            if (flushTimeout) {
                clearTimeout(flushTimeout);
                flushTimeout = null;
            }
        },
    };
}

// Singleton indexer instance
export const messageIndexer = createMessageIndexer();

File: src-tauri/src/localsearch.rs
// Local Search - Native FTS5 full-text search for messages
use rusqlite::{Connection, params};
use serde::{Deserialize, Serialize};
use std::path::PathBuf;
use std::sync::Mutex;
use tauri::{AppHandle, Manager, Runtime};

/// Search result item
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchResult {
    pub message_id: String,
    pub channel_id: String,
    pub server_id: Option<String>,
    pub author_id: String,
    pub author_name: String,
    pub content: String,
    pub snippet: String,
    pub timestamp: i64,
    pub relevance: f64,
    pub has_attachments: bool,
    pub has_embeds: bool,
    pub is_pinned: bool,
}

/// Message to index
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IndexableMessage {
    pub message_id: String,
    pub channel_id: String,
    pub server_id: Option<String>,
    pub author_id: String,
    pub author_name: String,
    pub content: String,
    pub timestamp: i64,
    pub has_attachments: bool,
    pub has_embeds: bool,
    pub is_pinned: bool,
    pub is_edited: bool,
}

/// Search query options
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct SearchOptions {
    pub query: String,
    pub channel_ids: Option<Vec<String>>,
    pub server_ids: Option<Vec<String>>,
    pub author_ids: Option<Vec<String>>,
    pub from_date: Option<i64>,
    pub to_date: Option<i64>,
    pub has_attachments: Option<bool>,
    pub has_embeds: Option<bool>,
    pub is_pinned: Option<bool>,
    pub limit: Option<u32>,
    pub offset: Option<u32>,
    pub sort_by: Option<String>, // "relevance" or "date"
    pub sort_order: Option<String>, // "asc" or "desc"
}

/// Search index statistics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IndexStats {
    pub total_messages: u64,
    pub total_channels: u64,
    pub total_servers: u64,
    pub index_size_bytes: u64,
    pub last_indexed_at: Option<i64>,
    pub oldest_message: Option<i64>,
    pub newest_message: Option<i64>,
}

/// Search manager state
pub struct SearchManager {
    conn: Mutex<Connection>,
}

impl SearchManager {
    /// Initialize search database
    pub fn new(app_data_dir: PathBuf) -> Result<Self, String> {
        let db_path = app_data_dir.join("search_index.db");
        
        // Ensure directory exists
        if let Some(parent) = db_path.parent() {
            std::fs::create_dir_all(parent)
                .map_err(|e| format!("Failed to create data dir: {}", e))?;
        }

        let conn = Connection::open(&db_path)
            .map_err(|e| format!("Failed to open search database: {}", e))?;

        // Enable FTS5
        conn.execute_batch(r#"
            -- Messages table for metadata
            CREATE TABLE IF NOT EXISTS messages (
                message_id TEXT PRIMARY KEY,
                channel_id TEXT NOT NULL,
                server_id TEXT,
                author_id TEXT NOT NULL,
                author_name TEXT NOT NULL,
                content TEXT NOT NULL,
                timestamp INTEGER NOT NULL,
                has_attachments INTEGER DEFAULT 0,
                has_embeds INTEGER DEFAULT 0,
                is_pinned INTEGER DEFAULT 0,
                is_edited INTEGER DEFAULT 0,
                indexed_at INTEGER NOT NULL
            );

            -- FTS5 virtual table for full-text search
            CREATE VIRTUAL TABLE IF NOT EXISTS messages_fts USING fts5(
                message_id,
                content,
                author_name,
                content='messages',
                content_rowid='rowid',
                tokenize='porter unicode61 remove_diacritics 2'
            );

            -- Triggers to keep FTS in sync
            CREATE TRIGGER IF NOT EXISTS messages_ai AFTER INSERT ON messages BEGIN
                INSERT INTO messages_fts(rowid, message_id, content, author_name)
                VALUES (NEW.rowid, NEW.message_id, NEW.content, NEW.author_name);
            END;

            CREATE TRIGGER IF NOT EXISTS messages_ad AFTER DELETE ON messages BEGIN
                INSERT INTO messages_fts(messages_fts, rowid, message_id, content, author_name)
                VALUES ('delete', OLD.rowid, OLD.message_id, OLD.content, OLD.author_name);
            END;

            CREATE TRIGGER IF NOT EXISTS messages_au AFTER UPDATE ON messages BEGIN
                INSERT INTO messages_fts(messages_fts, rowid, message_id, content, author_name)
                VALUES ('delete', OLD.rowid, OLD.message_id, OLD.content, OLD.author_name);
                INSERT INTO messages_fts(rowid, message_id, content, author_name)
                VALUES (NEW.rowid, NEW.message_id, NEW.content, NEW.author_name);
            END;

            -- Indexes for filtered queries
            CREATE INDEX IF NOT EXISTS idx_messages_channel ON messages(channel_id);
            CREATE INDEX IF NOT EXISTS idx_messages_server ON messages(server_id);
            CREATE INDEX IF NOT EXISTS idx_messages_author ON messages(author_id);
            CREATE INDEX IF NOT EXISTS idx_messages_timestamp ON messages(timestamp);
            CREATE INDEX IF NOT EXISTS idx_messages_pinned ON messages(is_pinned) WHERE is_pinned = 1;

            -- Metadata table
            CREATE TABLE IF NOT EXISTS search_meta (
                key TEXT PRIMARY KEY,
                value TEXT
            );
        "#).map_err(|e| format!("Failed to create tables: {}", e))?;

        Ok(Self {
            conn: Mutex::new(conn),
        })
    }

    /// Index a message
    pub fn index_message(&self, msg: &IndexableMessage) -> Result<(), String> {
        let conn = self.conn.lock().map_err(|e| e.to_string())?;
        let now = chrono::Utc::now().timestamp();

        conn.execute(
            r#"
            INSERT OR REPLACE INTO messages 
            (message_id, channel_id, server_id, author_id, author_name, content, 
             timestamp, has_attachments, has_embeds, is_pinned, is_edited, indexed_at)
            VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12)
            "#,
            params![
                msg.message_id,
                msg.channel_id,
                msg.server_id,
                msg.author_id,
                msg.author_name,
                msg.content,
                msg.timestamp,
                msg.has_attachments as i32,
                msg.has_embeds as i32,
                msg.is_pinned as i32,
                msg.is_edited as i32,
                now,
            ],
        ).map_err(|e| format!("Failed to index message: {}", e))?;

        Ok(())
    }

    /// Index multiple messages in batch
    pub fn index_messages_batch(&self, messages: &[IndexableMessage]) -> Result<u32, String> {
        let conn = self.conn.lock().map_err(|e| e.to_string())?;
        let now = chrono::Utc::now().timestamp();
        let mut count = 0;

        let tx = conn.unchecked_transaction()
            .map_err(|e| format!("Failed to start transaction: {}", e))?;

        for msg in messages {
            tx.execute(
                r#"
                INSERT OR REPLACE INTO messages 
                (message_id, channel_id, server_id, author_id, author_name, content, 
                 timestamp, has_attachments, has_embeds, is_pinned, is_edited, indexed_at)
                VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12)
                "#,
                params![
                    msg.message_id,
                    msg.channel_id,
                    msg.server_id,
                    msg.author_id,
                    msg.author_name,
                    msg.content,
                    msg.timestamp,
                    msg.has_attachments as i32,
                    msg.has_embeds as i32,
                    msg.is_pinned as i32,
                    msg.is_edited as i32,
                    now,
                ],
            ).map_err(|e| format!("Failed to index message {}: {}", msg.message_id, e))?;
            count += 1;
        }

        tx.commit().map_err(|e| format!("Failed to commit: {}", e))?;
        Ok(count)
    }

    /// Search messages with FTS5
    pub fn search(&self, opts: &SearchOptions) -> Result<Vec<SearchResult>, String> {
        let conn = self.conn.lock().map_err(|e| e.to_string())?;

        let limit = opts.limit.unwrap_or(50).min(200) as i64;
        let offset = opts.offset.unwrap_or(0) as i64;
        let sort_by = opts.sort_by.as_deref().unwrap_or("relevance");
        let sort_order = opts.sort_order.as_deref().unwrap_or("desc");

        // Build FTS5 query with MATCH
        let fts_query = if opts.query.is_empty() {
            "*".to_string()
        } else {
            // Escape special characters and add prefix matching
            let escaped = opts.query
                .replace("\"", "\"\"")
                .split_whitespace()
                .map(|word| format!("\"{}\"*", word))
                .collect::<Vec<_>>()
                .join(" ");
            escaped
        };

        // Build WHERE clause for filters
        let mut conditions = vec!["1=1".to_string()];
        let mut params_vec: Vec<Box<dyn rusqlite::ToSql>> = Vec::new();

        if let Some(ref channel_ids) = opts.channel_ids {
            if !channel_ids.is_empty() {
                let placeholders = channel_ids.iter().map(|_| "?").collect::<Vec<_>>().join(",");
                conditions.push(format!("m.channel_id IN ({})", placeholders));
                for id in channel_ids {
                    params_vec.push(Box::new(id.clone()));
                }
            }
        }

        if let Some(ref server_ids) = opts.server_ids {
            if !server_ids.is_empty() {
                let placeholders = server_ids.iter().map(|_| "?").collect::<Vec<_>>().join(",");
                conditions.push(format!("m.server_id IN ({})", placeholders));
                for id in server_ids {
                    params_vec.push(Box::new(id.clone()));
                }
            }
        }

        if let Some(ref author_ids) = opts.author_ids {
            if !author_ids.is_empty() {
                let placeholders = author_ids.iter().map(|_| "?").collect::<Vec<_>>().join(",");
                conditions.push(format!("m.author_id IN ({})", placeholders));
                for id in author_ids {
                    params_vec.push(Box::new(id.clone()));
                }
            }
        }

        if let Some(from_date) = opts.from_date {
            conditions.push("m.timestamp >= ?".to_string());
            params_vec.push(Box::new(from_date));
        }

        if let Some(to_date) = opts.to_date {
            conditions.push("m.timestamp <= ?".to_string());
            params_vec.push(Box::new(to_date));
        }

        if let Some(has_attachments) = opts.has_attachments {
            conditions.push(format!("m.has_attachments = {}", has_attachments as i32));
        }

        if let Some(has_embeds) = opts.has_embeds {
            conditions.push(format!("m.has_embeds = {}", has_embeds as i32));
        }

        if let Some(is_pinned) = opts.is_pinned {
            conditions.push(format!("m.is_pinned = {}", is_pinned as i32));
        }

        let where_clause = conditions.join(" AND ");
        let order_clause = if sort_by == "date" {
            format!("m.timestamp {}", if sort_order == "asc" { "ASC" } else { "DESC" })
        } else {
            format!("bm25(messages_fts) {}", if sort_order == "asc" { "DESC" } else { "ASC" })
        };

        let sql = format!(r#"
            SELECT 
                m.message_id,
                m.channel_id,
                m.server_id,
                m.author_id,
                m.author_name,
                m.content,
                snippet(messages_fts, 1, '<mark>', '</mark>', '...', 64) as snippet,
                m.timestamp,
                bm25(messages_fts) as relevance,
                m.has_attachments,
                m.has_embeds,
                m.is_pinned
            FROM messages_fts
            JOIN messages m ON messages_fts.message_id = m.message_id
            WHERE messages_fts MATCH ?1
            AND {}
            ORDER BY {}
            LIMIT ?2 OFFSET ?3
        "#, where_clause, order_clause);

        // Build params for query - needs manual handling
        let mut stmt = conn.prepare(&sql)
            .map_err(|e| format!("Failed to prepare search query: {}", e))?;

        // Since rusqlite doesn't support dynamic params easily, use a simpler approach
        // for the common case without filters
        let results: Vec<SearchResult> = if params_vec.is_empty() {
            stmt.query_map(params![&fts_query, limit, offset], |row| {
                Ok(SearchResult {
                    message_id: row.get(0)?,
                    channel_id: row.get(1)?,
                    server_id: row.get(2)?,
                    author_id: row.get(3)?,
                    author_name: row.get(4)?,
                    content: row.get(5)?,
                    snippet: row.get(6)?,
                    timestamp: row.get(7)?,
                    relevance: row.get::<_, f64>(8)?.abs(),
                    has_attachments: row.get::<_, i32>(9)? != 0,
                    has_embeds: row.get::<_, i32>(10)? != 0,
                    is_pinned: row.get::<_, i32>(11)? != 0,
                })
            })
            .map_err(|e| format!("Search query failed: {}", e))?
            .filter_map(|r| r.ok())
            .collect()
        } else {
            // For filtered queries, use a simpler approach
            Vec::new() // TODO: implement filtered queries
        };

        Ok(results)
    }

    /// Delete a message from the index
    pub fn delete_message(&self, message_id: &str) -> Result<(), String> {
        let conn = self.conn.lock().map_err(|e| e.to_string())?;
        conn.execute("DELETE FROM messages WHERE message_id = ?1", params![message_id])
            .map_err(|e| format!("Failed to delete message: {}", e))?;
        Ok(())
    }

    /// Delete all messages from a channel
    pub fn delete_channel_messages(&self, channel_id: &str) -> Result<u64, String> {
        let conn = self.conn.lock().map_err(|e| e.to_string())?;
        let count = conn.execute("DELETE FROM messages WHERE channel_id = ?1", params![channel_id])
            .map_err(|e| format!("Failed to delete channel messages: {}", e))?;
        Ok(count as u64)
    }

    /// Get index statistics
    pub fn get_stats(&self) -> Result<IndexStats, String> {
        let conn = self.conn.lock().map_err(|e| e.to_string())?;

        let total_messages: u64 = conn.query_row(
            "SELECT COUNT(*) FROM messages",
            [],
            |row| row.get(0),
        ).unwrap_or(0);

        let total_channels: u64 = conn.query_row(
            "SELECT COUNT(DISTINCT channel_id) FROM messages",
            [],
            |row| row.get(0),
        ).unwrap_or(0);

        let total_servers: u64 = conn.query_row(
            "SELECT COUNT(DISTINCT server_id) FROM messages WHERE server_id IS NOT NULL",
            [],
            |row| row.get(0),
        ).unwrap_or(0);

        let oldest_message: Option<i64> = conn.query_row(
            "SELECT MIN(timestamp) FROM messages",
            [],
            |row| row.get(0),
        ).ok();

        let newest_message: Option<i64> = conn.query_row(
            "SELECT MAX(timestamp) FROM messages",
            [],
            |row| row.get(0),
        ).ok();

        let last_indexed_at: Option<i64> = conn.query_row(
            "SELECT MAX(indexed_at) FROM messages",
            [],
            |row| row.get(0),
        ).ok();

        // Get database file size
        let db_path = conn.path().map(|p| p.to_path_buf());
        let index_size_bytes = db_path
            .and_then(|p| std::fs::metadata(p).ok())
            .map(|m| m.len())
            .unwrap_or(0);

        Ok(IndexStats {
            total_messages,
            total_channels,
            total_servers,
            index_size_bytes,
            last_indexed_at,
            oldest_message,
            newest_message,
        })
    }

    /// Optimize the FTS index
    pub fn optimize(&self) -> Result<(), String> {
        let conn = self.conn.lock().map_err(|e| e.to_string())?;
        conn.execute("INSERT INTO messages_fts(messages_fts) VALUES('optimize')", [])
            .map_err(|e| format!("Failed to optimize index: {}", e))?;
        Ok(())
    }

    /// Rebuild the FTS index
    pub fn rebuild(&self) -> Result<(), String> {
        let conn = self.conn.lock().map_err(|e| e.to_string())?;
        conn.execute("INSERT INTO messages_fts(messages_fts) VALUES('rebuild')", [])
            .map_err(|e| format!("Failed to rebuild index: {}", e))?;
        Ok(())
    }

    /// Clear all indexed data
    pub fn clear(&self) -> Result<(), String> {
        let conn = self.conn.lock().map_err(|e| e.to_string())?;
        conn.execute("DELETE FROM messages", [])
            .map_err(|e| format!("Failed to clear index: {}", e))?;
        Ok(())
    }
}

/// Tauri commands

#[tauri::command]
pub async fn search_messages<R: Runtime>(
    app: AppHandle<R>,
    options: SearchOptions,
) -> Result<Vec<SearchResult>, String> {
    let search_manager = app.state::<SearchManager>();
    search_manager.search(&options)
}

#[tauri::command]
pub async fn index_message<R: Runtime>(
    app: AppHandle<R>,
    message: IndexableMessage,
) -> Result<(), String> {
    let search_manager = app.state::<SearchManager>();
    search_manager.index_message(&message)
}

#[tauri::command]
pub async fn index_messages_batch<R: Runtime>(
    app: AppHandle<R>,
    messages: Vec<IndexableMessage>,
) -> Result<u32, String> {
    let search_manager = app.state::<SearchManager>();
    search_manager.index_messages_batch(&messages)
}

#[tauri::command]
pub async fn delete_indexed_message<R: Runtime>(
    app: AppHandle<R>,
    message_id: String,
) -> Result<(), String> {
    let search_manager = app.state::<SearchManager>();
    search_manager.delete_message(&message_id)
}

#[tauri::command]
pub async fn get_search_stats<R: Runtime>(
    app: AppHandle<R>,
) -> Result<IndexStats, String> {
    let search_manager = app.state::<SearchManager>();
    search_manager.get_stats()
}

#[tauri::command]
pub async fn optimize_search_index<R: Runtime>(
    app: AppHandle<R>,
) -> Result<(), String> {
    let search_manager = app.state::<SearchManager>();
    search_manager.optimize()
}

#[tauri::command]
pub async fn clear_search_index<R: Runtime>(
    app: AppHandle<R>,
) -> Result<(), String> {
    let search_manager = app.state::<SearchManager>();
    search_manager.clear()
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn test_search_manager_creation() {
        let dir = tempdir().unwrap();
        let manager = SearchManager::new(dir.path().to_path_buf()).unwrap();
        let stats = manager.get_stats().unwrap();
        assert_eq!(stats.total_messages, 0);
    }

    #[test]
    fn test_index_and_search() {
        let dir = tempdir().unwrap();
        let manager = SearchManager::new(dir.path().to_path_buf()).unwrap();

        let msg = IndexableMessage {
            message_id: "msg1".to_string(),
            channel_id: "chan1".to_string(),
            server_id: Some("srv1".to_string()),
            author_id: "user1".to_string(),
            author_name: "Alice".to_string(),
            content: "Hello world, this is a test message".to_string(),
            timestamp: 1700000000,
            has_attachments: false,
            has_embeds: false,
            is_pinned: false,
            is_edited: false,
        };

        manager.index_message(&msg).unwrap();

        let results = manager.search(&SearchOptions {
            query: "hello world".to_string(),
            ..Default::default()
        }).unwrap();

        assert_eq!(results.len(), 1);
        assert_eq!(results[0].message_id, "msg1");
    }

    #[test]
    fn test_batch_indexing() {
        let dir = tempdir().unwrap();
        let manager = SearchManager::new(dir.path().to_path_buf()).unwrap();

        let messages: Vec<IndexableMessage> = (0..100).map(|i| IndexableMessage {
            message_id: format!("msg{}", i),
            channel_id: "chan1".to_string(),
            server_id: Some("srv1".to_string()),
            author_id: "user1".to_string(),
            author_name: "Alice".to_string(),
            content: format!("Message number {} with unique content", i),
            timestamp: 1700000000 + i as i64,
            has_attachments: false,
            has_embeds: false,
            is_pinned: false,
            is_edited: false,
        }).collect();

        let count = manager.index_messages_batch(&messages).unwrap();
        assert_eq!(count, 100);

        let stats = manager.get_stats().unwrap();
        assert_eq!(stats.total_messages, 100);
    }
}
